### ISSUE TITLE
OpenOptions doesn't work when append=true and truncate=true on unix platforms

### ISSUE THREAD

**Issue by alygin** [rust-lang/rust#34347]
```rust
	let mut file = OpenOptions::new()
                            .write(true)
                            .append(true)
                            .truncate(true)
                            .create(true)
                            .open("some_file.txt").unwrap();
```
This code causes panic on mac os:
`thread '<main>' panicked at 'called 'Result::unwrap()' on an 'Err' value: Error { repr: Os { code: 22, message: "Invalid argument" } }', ../src/libcore/result.rs:746
`

Changing `append(true)` to `append(false)` fixes the problem and the resulting `File` behaves as expected. But it looks like an error, there's nothing wrong in setting both `append` and `truncate` to `true` simultaneously.

**Comment by sfackler** [comment]
The responsible check is here: https://github.com/rust-lang/rust/blob/master/src/libstd/sys/unix/fs.rs#L382.

**Comment by sfackler** [comment]
This will error out for all unixy platforms, not just OSX, btw.

**Comment by tbu-** [comment]
`append` and `truncate` sound mutually exclusive.

**Comment by alygin** [comment]
I agree, one can say that there's something fishy about `append=true, truncate=true` and such combination must cause panic. But then there must be a clear and documented rule that any ambiguity is treated as an error. There is no such rule now and according to the code Steven pointed to, the combination `append=true, truncate=true, create_new=false` isn't allowed, while even more ambiguous `append=true, truncate=true, create_new=true` is acceptable, as well as totally wrong `write=false, append=true`.

I see three possible solutions here:
1. Panic in any case of a suspicion that the developer doesn't understand what he or she wants or cannot clearly and unambiguously explain it in code.
2. Ignore options that can be ignored without changing result (like `truncate=true` can be ignored in combination with `create_new=true`) and panic in any case of suspicion that the developer made a mistake (like in our case with `append=true, truncate=true`).
3. Ignore all the combinations that don't have obvious unambiguity.

Solution 1 looks like the Rust-way, but there's a good chance that it will break some code. Solution 2 looks like the golden mean here for me. Both 2 and 3 need clear documentation on the allowed combinations.

Unfortunately, the current implementation has no clear rules on the issue. At least none is documented or can be guessed from the source code.

Any thoughts?

**Comment by nagisa** [comment]
See https://github.com/rust-lang/rfcs/blob/master/text/1252-open-options.md

**Comment by nagisa** [comment]
> even more ambiguous append=true, truncate=true, create_new=true is acceptable

It is not ambiguous. `create_new` is equivalent to creating a new file and `truncate=true` can be simply ignored. Using `append=true` does not make `truncate` invalid because it is no-op in the first place.

> write=false, append=true

Both seem like somewhat mutually exclusive options, but I see how it could be confusing. See especially the implementation detail aside in [this section](https://github.com/rust-lang/rfcs/blob/master/text/1252-open-options.md#append-mode).

**Comment by alygin** [comment]
Thank you for sharing the link, Simonas! That's what I was looking for but somehow failed to find. The document fully clarifies the current implementation, though some things might seem a bit confusing after switching from other languages. More informative error messages in `OpenOptions` would surely help, but that's not the subject of this issue. It might be closed, I think.

**Comment by niamster** [comment]
It's a valid combination of options on unix, why it is considered as invalid?
This is responsibility of OS to check the valid `open` flags.
